### PR TITLE
[TIR] CSE-TIR Pass - More deterministic behavior

### PR DESCRIPTION
**Pull request by AndrewZhaoLuo**
Running the CSE-TIR pass led to some variation in lowering to TIR from the same schedule multiple times. I tracked this down to iterating over an unordered map. I think having consistency in lowering to TIR is a desirable property to have so I made the offending code a little better in this regard.

Example test script:
https://github.com/AndrewZhaoLuo/TVM-Sandbox/blob/19284ddbd6bb28af61c0c2aa8bb334c5c53731a7/tir/test_inconsistent_tir_lowering.py#L1 

^--- Running the above off main produces different hashes every time, running after this diff it produces the same hash.

There may be more examples of non-determinism in CSE-TIR but this is what I found for now 👀

### CODE REVIEW

**Comment by AndrewZhaoLuo** [comment]
cc some folks from https://github.com/apache/tvm/pull/9482: @masahi @FranckQC @mbs-octoml @jroesch

File: src/tir/transforms/common_subexpr_elim_tools.cc [hunk 1]
```diff
@@ -748,8 +748,26 @@ std::vector<std::pair<PrimExpr, size_t>> SyntacticToSemanticComputations(
...
+    std::stringstream a_stream;
+    std::stringstream b_stream;
+    a_stream << a;
+    b_stream << b;
+    return a_stream.str().compare(b_stream.str()) < 0;
+  });
+
+  // For each element in the hashtable
+  for (PrimExpr key : table_keys) {
+    size_t value = table.find(key)->second;
```
Comment (masahi) [inline src/tir/transforms/common_subexpr_elim_tools.cc:768]: I think it is better to sort and put the sorted key + count pair into a vector before calling `SyntacticToSemanticComputations`. That way you can remove this `table.find(key)` and `SyntacticToSemanticComputations` becomes `vector -> vector` transform (which is cleaner imo).

Comment (AndrewZhaoLuo) [reply src/tir/transforms/common_subexpr_elim_tools.cc]: Done

File: src/tir/transforms/common_subexpr_elim_tools.cc [hunk 2]
```diff
@@ -743,13 +743,30 @@ bool EquivalentTerms(const PrimExpr& a, const PrimExpr& b) {
...
   result.reserve(table.size());
 
-  // For each element in the hashtable
+  // Traverse through map in a sorted order on keys to maintain deterministic behavior
+  // We do this by comparing the string repr of each PrimExpr to get a determinstic ordering
+  std::vector<std::pair<PrimExpr, size_t>> sorted_map_items;
+  sorted_map_items.reserve(table.size());
   for (auto elem : table) {
+    sorted_map_items.push_back(elem);
+  }
```
Comment (masahi) [inline src/tir/transforms/common_subexpr_elim_tools.cc:758]: I think you can initialize directly by `sorted_map_items(table.begin(), table.end())`. Can you try it?

**Comment by masahi** [comment]
@AndrewZhaoLuo maybe we can also solve this by changing the comparison functor in 
https://github.com/apache/tvm/blob/09f7be28a1ed7e629ddc0c9fcf19d4feace05dd4/src/tir/transforms/common_subexpr_elim_tools.h#L54

File: src/tir/transforms/common_subexpr_elim_tools.cc (hunk 2 above)
Comment (AndrewZhaoLuo) [reply src/tir/transforms/common_subexpr_elim_tools.cc]: Ah yes, thanks.

**Comment by AndrewZhaoLuo** [comment]
Hmm, so the unordered map doesn't have an ordering so all it needs is some way of determining equality of elements. 

An alternative would be `std::map` which does have a notion of ordering (and you can define a comparator). However, it's performance characteristics are different (usually self-balancing binary tree so O(log n) lookup instead of O(~1)). 

In general though yeah, we need to be on the lookout for iterating over unordered collections cause it will be non-determinstic. I'll try refactoring this sometime later, though need to read more about performance (it's already kind of slow when processing a statement with 100 common subexpressions).

**Comment by masahi** [comment]
> Hmm, so the unordered map doesn't have an ordering

The actual order doesn't matter, does it? `unordered_map` doesn't store elements in a sorted order, but it does have some ordering internally determined by the comparison functor. And what we need is just a fixed order, not necessarily the sorted order, so I think `unordered map` should do the job.

**Comment by masahi** [comment]
ah sorry, I was confused with something, "but it does have some ordering internally determined by the comparison functor" is not correct.

**Comment by masahi** [comment]
@AndrewZhaoLuo I think we can merge this as is for now. The CI failed on the mac os, probably need to kick another job.

**Review by masahi — APPROVED** [review]

**Comment by FranckQC** [comment]
Hi,

Sorry I couldn't discuss this with you last Thursday/Friday. Thanks for improving the CSE!

Yes, indeed, iterating through the hash table will not necessary lead to accessing its elements always in the same order. That's because the iteration order depends on the hashes (and we use `structural_hash` for that), which themselves are based on pointers, so different run will give different addresses, which will give different hashes, and thus a different order in which elements are accessed.

I guess it is not an issue as far as the CSE is concerned, because after producing this hashtable, we use it to produce a vector of semantic entities (which are pairs <expr, count>).
The main loop of the CSE will later iterate through the elements of this vector.
But before that, this vector gets sorted by the size (called "complexity") of the expression (the first component).

So the order of the end result (the vector) is non-deterministic only for elements with the same size / complexity, for which it does not matter (again, only for the CSE) in which order we common them. (i.e. it should not add or remove opportunities for commoning more later). Said differently, this non-determinism should appear for orthogonal choices.
I guess as in:
```
Mem[i1] = (a+b)+ (c+d);
Mem[i2] = a+b;               // We can common this one before or after
Mem[i3] = c+d;               // this one, it does not matter for the CSE
Mem[i4] = (a+b) + (c+d) + e;
```

for which we could produce: 

```
let cse_var_2 = a+b in
let cse_var_3 = c+d in
let cse_var_1 = cse_var_2 + cse_var_3 in
Mem[i1] = cse_var_1;
Mem[i2] = cse_var_2 in
Mem[i3] = cse_var_3 in
Mem[i4] = cse_var_1
```

Or the alternative version were (c+d) gets introduced before (a+b).

I did not think that having this level of non-determinism would be an issue, but you're clearly right that its better to always produce exactly the same resulting TIR code, that just makes testing easier.
Your way of making it deterministic (by first transitioning the hashtable into an array, which you then sort using the actual syntax that the expressions represent) is perfectly fine to me.

According to the doc about `structural_hash` ( here : https://tvm.apache.org/docs/reference/api/python/ir.html ), it seems an alternative could have been to set `map_free_vars` to true, as that's the reason the hashes use pointers. If that was the only place were addresses are used in the hashes, it should make everything deterministic too. I'll see if that works too, out of curiosity.

Many thanks for having spotted that and for having fixed it!

Kind regards,

Franck

**Pull request merged**